### PR TITLE
Load GitHub users list at startup for improved authentication performance

### PR DESCRIPTION
**Pull request by rbren**
This PR implements loading the GitHub users list at startup instead of during each authentication request. This change improves the performance of the authentication process.

### CODE REVIEW

File: openhands/server/listen.py [hunk 1]
```diff
@@ -68,10 +68,40 @@
...
+    waitlist = os.getenv('GITHUB_USER_LIST_FILE')
+    if waitlist and os.path.exists(waitlist):
+        with open(waitlist, 'r') as f:
+            GITHUB_USER_LIST = [line.strip() for line in f if line.strip()]
+    else:
+        logger.warning("GitHub user list file not found or not specified.")
+
+GITHUB_CLIENT_ID = os.getenv('GITHUB_CLIENT_ID', '').strip()
+GITHUB_CLIENT_SECRET = os.getenv('GITHUB_CLIENT_SECRET', '').strip()
+
```
Comment (rbren) [inline openhands/server/listen.py:86]:
```suggestion
```

File: openhands/server/listen.py [hunk 2]
```diff
@@ -68,10 +68,40 @@
...
+app = FastAPI(lifespan=lifespan)
+app.add_middleware(
+    CORSMiddleware,
+    allow_origins=['http://localhost:3001', 'http://127.0.0.1:3001'],
+    allow_credentials=True,
+    allow_methods=['*'],
+    allow_headers=['*'],
+)
+
```
Comment (rbren) [inline openhands/server/listen.py:104]:
```suggestion
```

File: openhands/server/listen.py [hunk 3]
```diff
@@ -68,10 +68,40 @@
...
+# New function to load the user list
+def load_github_user_list():
+    global GITHUB_USER_LIST
+    waitlist = os.getenv('GITHUB_USER_LIST_FILE')
+    if waitlist and os.path.exists(waitlist):
+        with open(waitlist, 'r') as f:
+            GITHUB_USER_LIST = [line.strip() for line in f if line.strip()]
+    else:
+        logger.warning("GitHub user list file not found or not specified.")
+
```
Comment (rbren) [inline openhands/server/listen.py:83]:
```suggestion
load_github_user_list()
```

File: openhands/server/listen.py [hunk 4]
```diff
@@ -833,23 +846,17 @@ class User(BaseModel):
...
+    if GITHUB_USER_LIST:
+        if user is None or user.login not in GITHUB_USER_LIST:
             return JSONResponse(
-                status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
-                content={'error': 'Waitlist file not found'},
+                status_code=status.HTTP_403_FORBIDDEN,
+                content={'error': 'User not on waitlist'},
             )
+    else:
+        logger.warning("No GitHub user list available for authentication.")
```
Comment (rbren) [inline openhands/server/listen.py:859]:
```suggestion
```

File: openhands/server/listen.py [hunk 5]
```diff
@@ -68,6 +68,19 @@
 GITHUB_CLIENT_ID = os.getenv('GITHUB_CLIENT_ID', '').strip()
 GITHUB_CLIENT_SECRET = os.getenv('GITHUB_CLIENT_SECRET', '').strip()
 
+# New global variable to store the user list
+GITHUB_USER_LIST = []
```
Comment (rbren) [inline openhands/server/listen.py:72]:
```suggestion
GITHUB_USER_LIST = None
```

File: openhands/server/listen.py [hunk 6]
```diff
@@ -68,6 +68,19 @@
...
 GITHUB_CLIENT_SECRET = os.getenv('GITHUB_CLIENT_SECRET', '').strip()
 
+# New global variable to store the user list
+GITHUB_USER_LIST = None
+
+# New function to load the user list
+def load_github_user_list():
+    global GITHUB_USER_LIST
+    waitlist = os.getenv('GITHUB_USER_LIST_FILE')
+    if waitlist and os.path.exists(waitlist):
```
Comment (rbren) [inline openhands/server/listen.py:78]:
```suggestion
    if waitlist:
```

File: openhands/server/listen.py [hunk 7]
```diff
@@ -68,6 +68,19 @@
...
+# New function to load the user list
+def load_github_user_list():
+    global GITHUB_USER_LIST
+    waitlist = os.getenv('GITHUB_USER_LIST_FILE')
+    if waitlist:
+        with open(waitlist, 'r') as f:
+            GITHUB_USER_LIST = [line.strip() for line in f if line.strip()]
+    else:
+        logger.warning("GitHub user list file not found or not specified.")
+load_github_user_list()
```
Comment (rbren) [inline openhands/server/listen.py:83]:
```suggestion
```

File: openhands/server/listen.py [hunk 8]
```diff
@@ -866,3 +868,19 @@ async def get_response(self, path: str, scope):
...
+            return await super().get_response(path, scope)
+        except Exception:
+            # FIXME: just making this HTTPException doesn't work for some reason
+            return await super().get_response('index.html', scope)
+
+
+app.mount('/', SPAStaticFiles(directory='./frontend/build', html=True), name='dist')
+
+
+
```
Comment (rbren) [inline openhands/server/listen.py:886]:
```suggestion
```

File: openhands/server/listen.py [hunk 9]
```diff
@@ -68,6 +68,16 @@
...
+# New global variable to store the user list
+GITHUB_USER_LIST = None
+
+# New function to load the user list
+def load_github_user_list():
+    global GITHUB_USER_LIST
+    waitlist = os.getenv('GITHUB_USER_LIST_FILE')
+    if waitlist:
+        with open(waitlist, 'r') as f:
+            GITHUB_USER_LIST = [line.strip() for line in f if line.strip()]
```
Comment (rbren) [inline openhands/server/listen.py:81]: load_github_user_list()

Comment (rbren) [reply openhands/server/listen.py:81]:
```suggestion
load_github_user_list()
```

**Review by amanape — APPROVED** [review]
LGTM

You can also utilize [FastAPI's lifespan function](https://fastapi.tiangolo.com/advanced/events/#lifespan-function) to set the data before the application starts

E.g.,
```py
GITHUB_USER_LIST = None


@asynccontextmanager
async def lifespan(app: FastAPI):
    GITHUB_USER_LIST = get_waitlist_users()
    yield
    # Clean up (if applicable)


app = FastAPI(lifespan=lifespan)
```

>The first part of the function, before the `yield`, will be executed **before** the application starts.
And the part after the `yield` will be executed **after** the application has finished.

**Comment by rbren** [comment]
Going to merge this but lifespan looks neat

**Pull request merged**